### PR TITLE
Mini-batch training for recursive neural networks example

### LINKED ISSUE #2033: Mini-batch learning for recursive neural networks
Bowman[1] described mini-batch training technique for recursive neural networks. We need to implement two functions:

1. `thin_stack_get(s, i)`: Returns `s[range(len(i)), i]`
2. `thin_stack_assign(s, i, x)`: Returns a new stack that satisfies `s[range(len(i)), i] = x`

Both functions need to share a stack array and its gradient array.

[1] S. R. Bowman, et.al. A Fast Unified Model for Parsing and Sentence Understanding. ACL 2016.
http://nlp.stanford.edu/blog/hybrid-tree-sequence-neural-networks-with-spinn/

### PR DESCRIPTION
**Pull request by unnonouno**
I implemented thin-stack technique introduced by Bowman [1] as two functions `thin_stack_set` and `thin_stack_get`.

- `thin_stack_set(s, i, x)` set stack elements. It behaves like `s[range(len(i)), i] = x` and return new `s`, where `s` stores a variable map of stacks, `i` stores indexes, and `x` stores value to assign.
- `thin_stack_get(s, i)` get stack elements. It behaves like `s[range(len(i)), i]`.
- Both functions **reuse ndarrays** in both forward and backward process. So, users need to use these method very carefully because after they called a function an ndarray in a variable will be changed.

And I made a mini-batch training for recursive neural networks example using thin-stack. It is about 10 times faster than the original. Note that it is slower to converge with mini-batch training.

fix #2033

1: http://nlp.stanford.edu/pubs/bowman2016spinn.pdf

### CODE REVIEW

**Comment by sleepinyourhat** [comment]
Wow! This is quite exciting to see. Could you say more about the speed comparison?

**Comment by unnonouno** [comment]
We implemented a simple recursive neural network example for Stanford sentiment treebank. We first implemented it with simple recursive call, and it requires 400 seconds/epoch. And I implemented it with thin-stack technique, and it requires 50 second/epoch when mini-batch size is 25 (about x8 faster). When mini-batch size is 100, the speed is 16 second/epoch (about x25 faster). They are results of CPU.
With GPU, calculation is more and more faster, but data translation is now bottle neck. Maybe unit size=30 is too small.

**Comment by beam2d** [comment]
Sorry for being soooooo late. As you noted, the in-place behavior of `thin_stack_set` and `thin_stack_get` is important and require discussions because it can make backprop incorrect (even if forward prop is correct). Chainer does not have a way to automatically detect such a kind of bugs embedded by users.

Do you think it's still reasonable to provide these functions by Chainer itself? If so, we have to clearly note the way to ensure that the code will correctly compute the gradient. The following might be some safe rules:

1. For `thin_stack_set`, the input should not be used by any other functions.
2. For both `thin_stack_set` and `thin_stack_get`, the output should be used only once.

Under these restrictions, I think the gradient will be correctly computed.

I will review the whole diff including example code.

**Comment by beam2d** [comment]
We had offline discussions and concluded that at least once we will put the functions to example directory with tests. We need a document that explains the task, method, and implementation. It might be good to ask another member to write the documentation after sharing the details offline.

**Comment by niboshi** [comment]
@unnonouno 
Test is failing.
Could you merge master?

**Comment by unnonouno** [comment]
merged

**Comment by keisuke-umezawa** [comment]
このPR、exampleも付け加えているPRなのですが、説明をもっと書かないと実装を理解できないのではという話になっているのですが、興味ありますか？


3 件の返信
Keisuke Umezawa [2ヶ月前]
はい、是非やってみたいです。あとでPRの内容確認させてください。


Yuya Unno [2ヶ月前]
出た本もっていれば、実装の実装に書かれています（書きました）


Keisuke Umezawa [2ヶ月前]
出た本持ってます！確認しますね。

**Comment by beam2d** [comment]
OK. I will merge this PR, so then let's write the document!

**Comment by keisuke-umezawa** [comment]
@beam2d Thanks! I will do it.
